lib/rules/no-shadow.js

```diff
@@ -57,6 +67,70 @@ module.exports = {
             allow: (context.options[0] && context.options[0].allow) || []
         };
 
+        /**
+         * @callback CheckNodeCallback
+         * @param {ASTNode} node A node to check.
+         */
+
+        /**
+         * Traverses AST from a given node to find a node that verifies checkNodeCB condition.
+         * @param {ASTNode} node a node to get.
+         * @param {CheckNodeCallback} checkNodeCB a callback that checks whether or not the node verifies its condition or not.
+         * @returns {ASTNode} the node that verifies checkNodeCB condition.
+         */
+        function getNode(node, checkNodeCB) {
+            let currentNode = node;
+
+            while (currentNode && !checkNodeCB(currentNode)) {
+                currentNode = currentNode.parent;
+            }
+            return currentNode;
+        }
+
+        /**
+         * Checks if a variable and a shadowedVariable have the same init pattern node parent.
+         * @param {Object} variable a variable to check.
+         * @param {Object} shadowedVariable a shadowedVariable to check.
+         * @returns {boolean} Whether or not the variable and the shadowedVariable have the same init pattern node parent.
+         */
+        function areOnInitPatternNode(variable, shadowedVariable) {
+            const outerDef = shadowedVariable.defs[0];
+            const innerDef = variable.defs[0];
+
+            const callExpressionNode = getNode(
+                innerDef.node,
+                node => node.type === "CallExpression"
+            );
+
+            if (callExpressionNode) {
+                const initNode1 = getNode(
```

Comment:
It’s not clear what initNode1 and initNode2 represent. Can we rename them to better represent what they mean?